src/DurableTask.Netherite/StorageLayer/Faster/AzureBlobs/AzureStorageDevice.cs

```diff
@@ -603,14 +623,58 @@ async Task WriteToBlobAsync(BlobEntry blobEntry, IntPtr sourceAddress, long dest
                 await this.SingleWriterSemaphore.WaitAsync();
             }
 
-            long offset = 0;
-            while (numBytesToWrite > 0)
+            bool IsFaulted = false;
+
+            try
             {
-                var length = Math.Min(numBytesToWrite, MAX_UPLOAD_SIZE);
-                await this.WritePortionToBlobUnsafeAsync(blobEntry, sourceAddress, destinationAddress, offset, length, id).ConfigureAwait(false);
-                numBytesToWrite -= length;
-                offset += length;
+                long offset = 0;
+                while (numBytesToWrite > 0)
+                {
+                    var length = Math.Min(numBytesToWrite, MAX_UPLOAD_SIZE);
+                    await this.WritePortionToBlobUnsafeAsync(blobEntry, sourceAddress, destinationAddress, offset, length, id).ConfigureAwait(false);
+                    numBytesToWrite -= length;
+                    offset += length;
+                }
+            }
+            catch (Exception e)
+            {
+                IsFaulted = true;
+
+                // details about failed storage accesses are already traced by the error handler, but we still create another trace here
+                // to help us debug situations where exceptions are thrown in other places, and to clarify the execution path
+                this.BlobManager.StorageTracer?.FasterStorageError($"{nameof(WriteToBlobAsync)} id={id} encountered exception", e);
             }
+            finally
+            {
+                if (this.underLease)
+                {
+                    this.SingleWriterSemaphore.Release();
+                }
+            }
+
+            if (this.pendingReadWriteOperations.TryRemove(id, out ReadWriteRequestInfo request))
+            {
+                try
+                {
+                    if (IsFaulted)
+                    {
+                        this.BlobManager?.StorageTracer?.FasterStorageProgress($"StorageOpReturned AzureStorageDevice.WriteAsync id={id} (Failure)");
+                        request.Callback(uint.MaxValue, request.NumBytes, request.Context);
+                    }
+                    else
+                    {
+                        this.BlobManager?.StorageTracer?.FasterStorageProgress($"StorageOpReturned AzureStorageDevice.WriteAsync id={id}");
+                        request.Callback(0, request.NumBytes, request.Context);
+                    }
+                    this.BlobManager?.StorageTracer?.FasterStorageProgress($"AzureStorageDevice.WriteAsync FASTER callback completed id={id}");
+                }
+                catch (Exception e)
+                {
+                    this.BlobManager.StorageTracer?.FasterStorageError($"{nameof(WriteToBlobAsync)} id={id} failed during FASTER completion callback", e);
+                }
```

Comment:
should we re-throw?

Reply:
There is no one to catch this. (see comment at bottom of method)